### PR TITLE
bugfix: S3C-1959 Sanity check for data deletion

### PR DESCRIPTION
**Pull request by rahulreddy**
## Description
Add sanity check for deletion of data keys
### Motivation and context

When metadata service is unstable it's possible that an object has successfully
completed composing, final object written to the destination bucket but metadata
service returned HTTP code 500. In this case, the server returns HTTP code 500 to the client and
if the client retries the complete mpu request, it tries to compose the object again but this
time accidentally deleting data keys from the previous request as it assumes that it is overwriting
an existing object where in reality it is deleting the very own keys of the object it is composing.
This check ensures that the current object being composed does not
accidentally delete current keys.

### Other commits accompanying this PR

- test: process mpu parts
- improvement: avoid coverage use npm test instead
- bugfix: fix typo in test setup

### CODE REVIEW

**Comment by rahulreddy** [comment]
I suggest reviewing commit by commit for context.

File: lib/api/apiUtils/object/locationKeysSanityCheck.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+* @param {array} curr - list of keys to be used in composing current object
+* @returns {array} list of keys that can be deleted
+*/
+function locationKeysSanityCheck(prev, curr) {
+    // backwards compatibility check if object is of model version 2
+    if (typeof prev === 'string') {
+        return curr.every(v => v.key !== prev);
+    }
+    const keysList = prev.map(v => v.key);
+    return curr.every(v => !keysList.includes(v.key));
```
Comment (jonathan-gramain) [inline lib/api/apiUtils/object/locationKeysSanityCheck.js]: Looks like this line has an O(n^2) complexity in the number of parts, I fear it could cause long lock-ups of the process with 1-10K MPUs being completed on top of an existing 1-10K MPU, as the node process could loop up to 100M times without giving hand to the event loop.

I suggest to store all prev keys in a js object first instead of using `Array.map`, to benefit from an efficient lookup.

**Review by jonathan-gramain — COMMENTED** [review]
I'm good with the philosophy of the fix, just have a technical concern.

File: lib/api/apiUtils/object/locationKeysSanityCheck.js (hunk 1 above)
Comment (rahulreddy) [reply lib/api/apiUtils/object/locationKeysSanityCheck.js]: I debated it but I didn't think of the case where a 10K mpu could replace a 10K mpu.

File: lib/api/apiUtils/object/locationKeysSanityCheck.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* @param {array|string} prev - list of keys from the object being overwritten
+* @param {array} curr - list of keys to be used in composing current object
+* @returns {array} list of keys that can be deleted
+*/
+function locationKeysSanityCheck(prev, curr) {
+    // backwards compatibility check if object is of model version 2
+    if (typeof prev === 'string') {
+        return curr.every(v => v.key !== prev);
+    }
+    const keysList = prev.map(v => v.key);
```
Comment (bennettbuchanan) [inline lib/api/apiUtils/object/locationKeysSanityCheck.js]: I'm not sure if this case can occur, but we could make an optimization by returning `true` if `keysList.length === 0`.

Comment (rahulreddy) [reply lib/api/apiUtils/object/locationKeysSanityCheck.js]: I can add a check at the top.

**Comment by rahulreddy** [comment]
@bert-e reset

**Comment by rahulreddy** [comment]
@bert-e reset

File: lib/api/apiUtils/object/locationKeysSanityCheck.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+*/
+function locationKeysSanityCheck(prev, curr) {
+    if (!prev || prev.length === 0) {
+        return true;
+    }
+    // backwards compatibility check if object is of model version 2
+    if (typeof prev === 'string') {
+        return curr.every(v => v.key !== prev);
+    }
+    const keysList = {};
```
Comment (bennettbuchanan) [inline lib/api/apiUtils/object/locationKeysSanityCheck.js]: ~keysList~ keysHash

File: eve/main.yml [hunk 4]
```diff
@@ -140,8 +140,8 @@ stages:
           command: |
             set -ex
             mkdir -p $CIRCLE_TEST_REPORTS/unit
-            npm run unit_coverage
-            npm run unit_coverage_legacy_location
+            npm test
```
Comment (jonathan-gramain) [inline eve/main.yml:143]: Was this change intentional?

File: tests/unit/api/apiUtils/processMpuParts.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    });
+
+    [0, 2, 4].forEach(n => {
+        it(`should filter ${n} parts that are not used in complete mpu`,
+            () => {
+                const expected = _buildExpectedResult(n);
+                const result = validateAndFilterMpuParts(storedParts, jsonList,
+                    mpuOverviewKey, splitter, log);
+                // console.log(expected, result);
+                assert(expected, result);
```
Comment (bennettbuchanan) [inline tests/unit/api/apiUtils/processMpuParts.js]: Looks like this is meant to be `assert.deepStrictEqual`?

File: tests/unit/api/apiUtils/processMpuParts.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        splitter = null;
+    });
+
+    [0, 2, 4].forEach(n => {
+        it(`should filter ${n} parts that are not used in complete mpu`,
+            () => {
+                const expected = _buildExpectedResult(n);
+                const result = validateAndFilterMpuParts(storedParts, jsonList,
+                    mpuOverviewKey, splitter, log);
+                // console.log(expected, result);
```
Comment (bennettbuchanan) [inline tests/unit/api/apiUtils/processMpuParts.js]: Extra comment here.

Comment (jonathan-gramain) [inline tests/unit/api/apiUtils/processMpuParts.js]: debug comment to be removed

File: eve/main.yml (hunk 4 above)
Comment (rahulreddy) [reply eve/main.yml:143]: Yes, we removed this in 8.0, 8.1 but did not backport it.

**Comment by rahulreddy** [comment]
@bert-e reset

**Review by bennettbuchanan — APPROVED** [review]

**Review by jonathan-gramain — APPROVED** [review]

**Comment by rahulreddy** [comment]
@bert-e reset

**Comment by rahulreddy** [comment]
@bert-e approve
